Users management. Page made from scratch.

PiekosEmil opened the pull request:
Do poprawy.
JavaScript do osobnego pliku.
Nieobsugiwane parametry na stronie login, które czasami powodują błędy przy rejestracji lub wylogowaniu. Trzeba wtedy ponownie wpisywać dane logowania.

Pozostałe punkty zadania wykonane zgodnie z opisem.

PiekosEmil committed:
Do poprawy.
JavaScript do osobnego pliku.
Nieobsugiwane parametry na stronie login, które czasami powodują błędy przy rejestracji lub wylogowaniu.
Trzeba wtedy ponownie wpisywać dane logowania.

Pozostałe punkty zadania wykonane zgodnie z opisem.

marcinkunert commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,10 @@
+INSERT INTO user(username, password)
+VALUES ('RalphDemolka', '{bcrypt}$2a$12$y7v0CrAsIxK4K3wHBJrx8e2Opm6R9UyKDBqWHC61Ktc/BuygfJ2w2'), -- prostehaslo
+       ('Bobber', '{bcrypt}$2a$12$8LlJA4PEy7Eb.0H1OkmXmuJ6hzdIlFtMoDKfbysfooQuSGF2SVtD6'), -- trudniejszehaslo
+       ('TestowyUser', '{bcrypt}$2a$12$OFKFoJrQQv0GFA9gL7v4HOpp59N5aCB2/cZoxHVEDZrtALxsn.r6.'); -- testowehaslo
```
Zamiast komentarzy możesz użyć agorytmu noop (non-operative) i trzymać hasło w plaintext.
`{noop}prostehaslo`

PiekosEmil replied:
Tak początkowo zrobiłem, nie wiedząc, że Spring automatycznie decoduje hasła. Sądziłem, że trzeba jakoś ustalic algorytm dekodowania. Kiedy przekonałem się, że tak nie jest zakodowałem je, ponieważ w zadaniu jest informajca, że hasła mają byc szyfrowane dlatego zakodowałem je pierwszym lepszym encoderem z Google. Ale teraz kiedy piszę ten tekst dociera do mnie, że zadanie mogło dotyczyc haseł nowych użytkowników. :-)

marcinkunert commented on src/main/java/com/emilpiekos/usersmanager/controllers/AdminController.java:
```diff
@@ -0,0 +1,70 @@
+@Controller
+public class AdminController {
+
+    @Autowired
+    UsersService usersService;
+
+    @Autowired
+    UserRoleService userRoleService;
```
Zamiast wstrzykiwać zależości przez pola, lepiej wstrzykiwać przez konstruktor:

```
public AdminController(UsersService userService, UserRoleService userRoleService) {
...
```

PiekosEmil replied:
Tak też zastanawiałem się, czemu IntelliJ podpowiada, żeby nie używac, tej adnotacji. Początkowo wszędzie miałem konstruktory, ale zmyliły mnie rozwiązania na stackoverflow. :-D

marcinkunert commented on src/main/java/com/emilpiekos/usersmanager/controllers/UserController.java:
```diff
@@ -0,0 +1,59 @@
+
+    @GetMapping("/")
+    public String userPage(@RequestParam(required = false) String username, Model model) {
+        boolean isAdmin = SecurityContextHolder.getContext().getAuthentication()
+                .getAuthorities().stream().anyMatch(a -> a.getAuthority().equals("ROLE_ADMIN"));
+        String currentUsername = SecurityContextHolder.getContext().getAuthentication().getName();
+        model.addAttribute("currentUsername", currentUsername);
+        model.addAttribute("isAdmin", isAdmin);
```
Sprawdzanie w szablonie czy użytkownik jest adminem można obsłużyć dodatkową biblioteką: https://github.com/thymeleaf/thymeleaf-extras-springsecurity

Tutaj przykład możliwości (nie trzeba nic dodawać w kontrolerze):
```
<div sec:authorize="hasRole('ROLE_ADMIN')">
  This content is only shown to administrators.
</div>
```

marcinkunert reviewed (approved):
Jest zrobione bardzo dobrze. Dodałem kilka komentarzy co można by ewentualnie poprawić.

PiekosEmil committed:
Poprawka wyświetlania dodatkowego elementu <div> w zależności od przysługującego przywileju.

PiekosEmil committed:
Poprawka ograniczeń dostępu do treści administracyjnych.
Wydzielono kod JavaScript do osobnego pliku.
zamieniono wstrzykiwanie przez adnotacje @Autowired na wstrzykiwanie przez konstruktor.

PiekosEmil committed:
Poprawa CSS.
Dodanie dodatkowych pól do późniejszej walidacji.
Dodanie strony informującej o pomyślnej rejestracji.
Configuracja SecurityFilterChain dodano atrybut wyłączający request cache. Kiedy opcja była włączona, niektóre adresy były przekształcane na wadliwe.
